src/scicloj/02_numpy.clj

```diff
@@ -140,7 +159,94 @@ np.random.randint(0, 10, (3, 3))
 np.eye(3)
 ```"]
 
-;; N/A
+["We currently do not have a good approximation of this technique.  However,
+it is easy enough to compute an identity matrix using an outer product operation."]
+
+(defn ^:private slow-index
+  "Like range, but for tensors.  Enumerates the indicies."
+  [shape]
+  (let [space  (apply * shape)
+        rshape (reverse shape)
+        muls   (cons 1 (pop (vec (reductions * rshape))))]
+    (-> (mapv
+         (fn [idx]
+           (vec 
+            (reverse
+             (map (fn [p n]
+                    (-> idx
+                        (quot n)
+                        (mod p)))
+                  rshape
+                  muls))))
+         (range space))
+        dtype/->reader)))
+
+(defn index [args]
+  (if-let [f (and (< (count args) 15)
+                  (case (count args)
+                    0 nil
+                    1 (fn [a] [a])
+                    2 (fn [a b] [a b])
+                    3 (fn [a b c]
+                        [a b c])
+                    4 (fn [a b c d]
+                        [a b c d])
+                    5 (fn [a b c d e]
+                        [a b c d e])
+                    6 (fn [a b c d e f]
+                        [a b c d e f])
+                    7 (fn [a b c d e f g]
+                        [a b c d e f g])
+                    8 (fn [a b c d e f g h]
+                        [a b c d e f g h])
+                    9 (fn [a b c d e f g h i]
+                        [a b c d e f g h i])
+                    10 (fn [a b c d e f g h i j]
+                         [a b c d e f g h i j])
+                    11 (fn [a b c d e f g h i j k]
+                         [a b c d e f g h i j k])
+                    12 (fn [a b c d e f g h i j k l]
+                         [a b c d e f g h i j k l])
+                    13 (fn [a b c d e f g h i j k l m]
+                         [a b c d e f g h i j k l m])
+                    14 (fn [a b c d e f g h i j k l m n]
+                         [a b c d e f g h i j k l m n])
+                    15 (fn [a b c d e f g h i j k l m n o]
+                         [a b c d e f g h i j k l m n o])))]
+    (-> (dtt/compute-tensor (vec args) f)
+        dtype/->reader))
+  (slow-index (vec args)))
+
+(index [3 3])
+
+(defn outer-product
+  ;; Author: Chris Nuernberger
+  ;; https://github.com/scicloj/scicloj-data-science-handbook/pull/2#discussion_r548033202
+  [f a b]
+  (let [a-shape (dtype/shape a)
+        b-shape (dtype/shape b)
+        a-rdr   (dtype/->reader a)
+        b-rdr   (dtype/->reader b)
+        n-b     (.lsize b-rdr)
```

Comment:
Great question and I think in most places potentially including this one `count` is the right primitive to use.  One important difference is that `count` works in `:int32` space so the maximum count of anything is around 2GB.  `lsize` was added because there is on `size` or `count` member of java.util.List that works in `:int64` space.  All of the [get/set methods](https://github.com/cnuernber/dtype-next/blob/master/java/tech/v3/datatype/Buffer.java#L24) in dtype-next work in long space as does the `lsize` - long-size function.  This design allows us to, for instance, mmap files larger than 2GB which we is useful when you want to load really large datasets such as large arrow files.

Reply:
I think also count is slightly slower as it will eventually figure out the interface method to call but that runtime dispatch takes just a small bit of time.  I didn't mention it because in this case I think it is irrelevant but there are sections of the dtype-next code, specifically in the tensor dimensions/indexing system where it is very much relevant.